### PR TITLE
Closes #4666: Dynamically support JSON extra file from GeckoView

### LINKED ISSUE #4666: Support upcoming crash format changes
See https://bugzilla.mozilla.org/show_bug.cgi?id=1420363

### PR DESCRIPTION
**Pull request by rocketsroger**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 1]
```diff
@@ -244,7 +249,12 @@ class MozillaSocorroService(
     }
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal fun readExtrasFromFile(file: File): HashMap<String, String> {
+    internal fun gsonUnescape(string: String): String {
+        return string.replace("""\\\\""", "\\").replace("""\n""", "\n").replace("""\t""", "\t")
```
Comment (rocketsroger) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Strings return from GSON can only be un-escaped correctly using the triple-quoted raw string.  Not exactly sure why.

**Comment by rocketsroger** [comment]
This is in preparation of future GeckoView changes that will format extras file in a crash in JSON.  The test files are directly copied from the GeckoView unit tests so it should work properly when GeckoView change is submitted.

**Comment by rocketsroger** [comment]
I'll use Android JSON support to avoid adding the GSON dependency.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 2]
```diff
@@ -277,6 +286,35 @@ class MozillaSocorroService(
...
+                val input = fileReader.readLines().firstOrNull()
+                input ?: throw JSONException("failed to read json file")
+
+                val jsonObject = JSONObject(input)
+                for (key in jsonObject.keys()) {
+                    resultMap[key] = jsonUnescape(jsonObject.getString(key))
+                }
+            }
+        } catch (e: FileNotFoundException) {
+            Logger.error("failed to find extra file", e)
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: nit: We also need to catch `IOException` here.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 3]
```diff
@@ -277,6 +286,35 @@ class MozillaSocorroService(
...
 
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun readExtrasFromFile(file: File): HashMap<String, String> {
+        var resultMap = HashMap<String, String>()
+        var notJson = false
+
+        try {
+            FileReader(file).use { fileReader ->
+                val input = fileReader.readLines().firstOrNull()
+                input ?: throw JSONException("failed to read json file")
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: nit: You can also write this as one statement:
```kotlin
val input = fileReader.readLines().firstOrNull()
    ?: throw JSONException("failed to read json file")
```

Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Nice, thanks

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 2 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Thanks, I'll add it

**Comment by rocketsroger** [comment]
bors r=pocmo

**Pull request merged**